fastpass/src/main/scala/scala/meta/internal/fastpass/bazelbuild/ActionGraph.scala

```diff
@@ -0,0 +1,136 @@
+package scala.meta.internal.fastpass.bazelbuild
+
+import scala.collection.JavaConverters.asScalaBufferConverter
+import scala.collection.mutable.Buffer
+import scala.collection.mutable.HashSet
+
+import scala.meta.internal.fastpass.bazelbuild.AnalysisProtosV2._
+
+class ActionGraph(
+    labelToActions: Map[String, List[Action]],
+    idToDepSetOfFiles: Map[Int, DepSetOfFiles],
+    idToArtifact: Map[Int, Artifact],
+    idToPathFragment: Map[Int, PathFragment]
+) {
+
+  def outputsOf(label: String): List[Artifact] = {
+    labelToActions
+      .getOrElse(label, Nil)
+      .flatMap(_.getOutputIdsList().asScala)
+      .flatMap(x => idToArtifact.get(x))
+  }
+
+  def transitiveInputsOf(
+      label: String,
+      mnemonics: Set[String]
+  ): List[Artifact] = {
+    for {
+      action <- labelToActions.getOrElse(label, Nil)
+      if mnemonics.contains(action.getMnemonic())
```

Comment:
`mnemonics.get(action).map(transitiveInputsOf).getOrElse(Nil)`? I'm surprised that this actually compiles... does it desugar to just a `filter`? If so, it's fine if you leave it. On the other hand, the `for` comp can be in `Option` and the `getOrElse` could be in the end.